### ISSUE TITLE
Restore fails with error "no space left on device" if PVCs are 100% utilised during backup

### ISSUE THREAD

**Issue by nrb** [velero-io/velero#2812]
**What steps did you take and what happened:**
From an internal bug report:

Using Restic, restore is failing with error "no space left on device" if PVC was fully occupied by the user data while the backup is taken.

for example, PVC of size 1Gi and user data in the PVC is ~upto 100% which is of 1Gi

**What did you expect to happen:**

Velero restic restores should be successful, even if the PV is full.

**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

```
root@k8-master-277:~# velero restore describe restore001 --details
Name:         restore001
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  PartiallyFailed (run 'velero restore logs restore001' for more information)

Errors:
  Velero:   pod volume restore failed: error restoring volume: error creating .velero directory for done file: mkdir /host_pods/60880bf2-9d1c-47cf-ba8f-5f8db43b385b/volumes/kubernetes.io~csi/pvc-ee79e0f0-ed62-44f7-987e-16efca1d1cd5/mount/.velero: no space left on device
  Cluster:    <none>
  Namespaces: <none>
```

**Anything else you would like to add:**

Since this is related to creating the `.velero` donefile, the fix might be somewhat relate to #2722.

**Environment:**

- Velero version (use `velero version`): v1.4.0
- Cloud provider or hardware configuration: vSphere

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by nrb** [comment]
Scoping this for v1.6 currently; I'm unsure if the fix will be backport-able at this time.

**Comment by sauerburger** [comment]
Is there an update regarding #2722? The left-over .velero/ directory causes quite a few headaches for me.

**Comment by larssb** [comment]
Seeing this in Velero v1.15.2. But, could it really be the same cause of error? We're using the velero-csi-plugin to trigger Longhorn to backup via the snapshot-controller and then the Velero CSI snapshot data mover feature.

Most PVC's successfully restore on another cluster ( we're migrating ). However, for three volumes we see this `no space left on device` error.

```text
│   Completion Timestamp:  2025-02-03T15:33:46Z                                                                 │
│   Message:               data path restore failed: Failed to run data path service for DataDownload cluster-m │
│ anager-0302251437-20250203153312-bdfbz: Data path for data download cluster-manager-0302251437-20250203153312 │
│ -bdfbz failed: Failed to run kopia restore: Failed to copy snapshot data to the target: restore error: copy f │
│ ile: error creating file: cannot write data to file "/21962707-dda9-4b52-b8d1-195813bfbb8f/kafka-log0/__consu │
│ mer_offsets-1/00000000000000000000.index": write /21962707-dda9-4b52-b8d1-195813bfbb8f/kafka-log0/__consumer_ │
│ offsets-1/00000000000000000000.index: no space left on device 
```

How can we fix this/is there a fix or a workaround? Thanks

**Comment by larssb** [comment]
I guess I'll try `--write-sparse-files ` implemented to fix #6664 - then we'll see. Maybe that is even the general fix for this issue as well.

**Comment by weshayutin** [comment]
@kaovilai will propose documentation or fix here. 
https://issues.redhat.com/browse/OADP-5739

**Comment by larssb** [comment]
Yeah documentation is the likely fix here. Using `--write-sparse-files` counters this.

**Comment by kaovilai** [comment]
https://github.com/vmware-tanzu/velero/pull/9222/ complements and/or replace incremental error message + doc improvements at https://github.com/vmware-tanzu/velero/pull/9221

**Comment by reasonerjt** [comment]
@kaovilai 
Let me clarify, is it for tracking the issue when the restic/kopia successfully writes the content during restore but the `done` file is failed to be written?  Or does it also cover the case that restic/kopia fails to write all the content due to lack of resource?

I think these are 2 different problems.

**Comment by Lyndon-Li** [comment]
We are investigating Volume Only Restore https://github.com/vmware-tanzu/velero/issues/7345, and we have one proposal to use Velero Generic Restore for PVR, so that the restore-helper init-container is not required. Then this problem will be fixed.

**Comment by kaovilai** [comment]
> one proposal to use Velero Generic Restore for PVR, so that the restore-helper init-container is not required.

That would work for me. Curious if there is no done file then does the workflow support workload quiescing.

**Comment by kaovilai** [comment]
> Let me clarify, is it for tracking the issue when the restic/kopia successfully writes the content during restore but the done file is failed to be written? Or does it also cover the case that restic/kopia fails to write all the content due to lack of resource?

IMO this is for tracking all issues that arises from when disk is fully utilized. Which blocks done file writing.

If a disk that got backed up later have data added and FSB cannot restore all data because there is not enough space left from post-backup-added-data then that's a separate issue.

**Comment by kaovilai** [comment]
another idea from meeting: on restore bump pvc size by X% from backup pvc size

**Comment by sseago** [comment]
> another idea from meeting: on restore bump pvc size by X% from backup pvc size

This should only be done optionally at user request, and per PV. This won't be what users will expect or want for volumes that are not almost full and are being managed appropriately, but it might be a good option for restore for specific volumes that are full.

In addition, when looking at possible PVC modifications to the request capacity, it might be worth looking at the PV size as well. It's possible to have a PV that's larger than the PVC request size (since the PVC request is a minimum). If the PV is larger than the PVC request size, it's possible to have a restore failure, especially on storageclass change, since different storage classes have different minimum PV size change increments. So we could track the PV size as a PVC annotation at backup time, and then when restoring PVC, set the request size to be the maximum of the PVC request and the PV actual size (as recorded in the annotation). Another scenario where PV size could be larger -- perhaps much larger -- than the PVC request size is for statically-provisioned NFS PVs. For example, if create a 1GB PVC that is expected to be bound to one of the precreated unused NFS PVs, and those are all 10GB in size, then my 1GB PVC will get bound to a volume ten times larger. This means that the resulting mounted PVC could have 10x as much data as indicated by the PVC request size, so on restore it could get a PV that's only 1GB, causing failure.

**Comment by kaovilai** [comment]
For resizing PVC on restore (not necessarily related to done file/initContainer issue solved via #9270), user may be able to use existing resourceModifier functionality https://github.com/vmware-tanzu/velero/issues/3854#issuecomment-3303534417

**Comment by sseago** [comment]
> For resizing PVC on restore (not necessarily related to done file/initContainer issue solved via [#9270](https://github.com/vmware-tanzu/velero/issues/9270)), user may be able to use existing resourceModifier functionality [#3854 (comment)](https://github.com/vmware-tanzu/velero/issues/3854#issuecomment-3303534417)

It may be worth testing that this works, but this might be good enough as a workaround if we expect it to only be needed on rare occasions -- i.e. "restore failed, lets try again with the resource modifier".

**Comment by kaovilai** [comment]
No bandwidth right now dropping this issue from candidates. If anyone have ideas for solving and drop here before feature freeze can think about it. Open to anyone else taking over.
